### PR TITLE
[MoE] notebook for numerical verification

### PR DESCRIPTION
**Pull request by ZhiyuLi-goog**
# Description

In addition to the typo, looks like we are missing this [normalization](https://github.com/huggingface/transformers/blob/main/src/transformers/models/mixtral/modeling_mixtral.py#L643) in [maxtext](https://github.com/AI-Hypercomputer/maxtext/blob/main/MaxText/layers/linears.py#L546), which is to re-scale top_k_weights to a normalized one where their sum equals 1. I was able to match layer outputs after adding the normalization. This normalization won't affect training/inferencing since this is a constant term (per each token), which won't change softmax probability. But we can still add it for better alignment.

Update: 
fix typo and add normalization for top_k_weights was covered in these PRs #1100 & #1064
Just add numerical verification sheet in this PR.

# Tests

Tested locally.

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed.

### CODE REVIEW

**Comment by gobbleturk** [comment]
Thanks for the fix Zhiyu. I'll leave for @RissyRan to review

**Comment by RissyRan** [comment]
Thanks Zhiyu for the fix! I am fine to add this normalization if this makes us convenient to compare weights. Could you add this `top_k_weights /= top_k_weights.sum(-1, keepdims=True)` before this line `weights = self.reshape_and_update_weights(top_k_weights, top_k_indices)`?  It was not used anywhere before that.

**Comment by RissyRan** [comment]
I think we also need to add this normalization to somewhere [here]https://github.com/AI-Hypercomputer/maxtext/blob/061abd8a375b6b279eacf0ae867e05e4e08cc360/MaxText/layers/linears.py#L488) for dropping. `softmax_probs *= combined_expert_mask`

Have you checked the benchmark scores with and without this normalization? using correct ckpt on 8x22b or 8x7b.

**Comment by ZhiyuLi-goog** [comment]
> Thanks Zhiyu for the fix! I am fine to add this normalization if this makes us convenient to compare weights. Could you add this `top_k_weights /= top_k_weights.sum(-1, keepdims=True)` before this line `weights = self.reshape_and_update_weights(top_k_weights, top_k_indices)`? It was not used anywhere before that.

Done.

**Comment by ZhiyuLi-goog** [comment]
> I think we also need to add this normalization to somewhere [here]
> 
> https://github.com/AI-Hypercomputer/maxtext/blob/061abd8a375b6b279eacf0ae867e05e4e08cc360/MaxText/layers/linears.py#L488
> 
> ) for dropping. `softmax_probs *= combined_expert_mask`
> Have you checked the benchmark scores with and without this normalization? using correct ckpt on 8x22b or 8x7b.

Have checked the benchmark scores the results are the same with and without this normalization.

Additionally,added a numerical verification notebook.
Currently able get ~0.1 tolerance in logits for both megablox=True or False.

**Review by RissyRan — APPROVED** [review]
Thank you!

**Comment by ZhiyuLi-goog** [comment]
@gobbleturk could you review it at your convenience?

**Comment by ZhiyuLi-goog** [comment]
> @gobbleturk could you review it at your convenience?

Hi, @gobbleturk I need code owner's review as the final step of this PR, thank you!

**Review by gobbleturk — APPROVED** [review]
